### PR TITLE
v-2

### PR DESCRIPTION
**Pull request by ar4min**

### CODE REVIEW

File: src/Endpoint/PaymentGateway/RequestTypes/ReverseRequest.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    public function validate(): void
+    {
+        $this->validateMerchantId();
+        $this->validateAuthority();
+    }
+
+    private function validateMerchantId(): void
+    {
+        if ($this->merchantId === null || !preg_match('/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/', $this->merchantId)) {
```
Comment (MA7) [inline src/Endpoint/PaymentGateway/RequestTypes/ReverseRequest.php:24]: make public function for valid uuid

File: src/Endpoint/PaymentGateway/RequestTypes/ReverseRequest.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            throw new InvalidArgumentException('Invalid merchant_id format. It should be a valid UUID.');
+        }
+    }
+
+    private function validateAuthority(): void
+    {
+        if (!preg_match('/^A[0-9a-zA-Z]{35}$/', $this->authority)) {
```
Comment (MA7) [inline src/Endpoint/PaymentGateway/RequestTypes/ReverseRequest.php:31]: make public function for valid Authority

File: src/Endpoint/PaymentGateway/ResponseTypes/UnverifiedResponse.php [hunk 3]
```diff
@@ -6,10 +6,33 @@
...
 
-    public array $authorities;
+    public int $code;
+    public string $message;
+    public array $authorities = [];
+
+    public function __construct(array $data = null)
+    {
+        if ($data === null || !isset($data['code'], $data['message'])) {
+            $this->code = 0;
```
Comment (MA7) [inline src/Endpoint/PaymentGateway/ResponseTypes/UnverifiedResponse.php:18]: handle this
we don't have 0 response code

File: src/Options.php [hunk 4]
```diff
@@ -24,19 +23,25 @@ public function __construct(array $options = [])
...
-                'client_builder' => new ClientBuilder(),
-                'uri_factory' => Psr17FactoryDiscovery::findUriFactory(),
-                'base_url' => $this->arrayGet(getenv(), 'ZARINPAL_MERCHANT_URL', 'https://api.zarinpal.com'),
-                'merchant_id' => $this->arrayGet(getenv(), 'ZARINPAL_MERCHANT_KEY', 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx')
-            ]
-        );
+        $resolver->setDefaults([
+            'client_builder' => new ClientBuilder(),
+            'uri_factory' => Psr17FactoryDiscovery::findUriFactory(),
+            'base_url' => 'https://payment.zarinpal.com',
```
Comment (MA7) [inline src/Options.php:29]: add ENV ZARINPAL_MERCHANT_URL

File: tests/BaseTestCase.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    protected $options;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->mockClient = new MockClient();
+        $this->options = new Options([
+            'access_token' => 'mock-access-token',
+            'merchant_id' => '25fe4c36-66e4-11e9-a9e4-000c29344814',
```
Comment (MA7) [inline tests/BaseTestCase.php:21]: use https://fusionauth.io/dev-tools/uuid-generator for generate uuid

File: tests/PaymentGateway/PaymentGatewayTest.php [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        $zarinpal->setHttpClient($this->clientMock);
+
+        $this->gateway = new PaymentGateway($zarinpal);
+    }
+
+    public function testRequest()
+    {
+        $responseBody = [
+            'data' => [
+                'authority' => 'A00000000000000000000000000123456',
```
Comment (MA7) [inline tests/PaymentGateway/PaymentGatewayTest.php:36]: A0000000000000000000000000012b4A6

**Review by MA7 — CHANGES_REQUESTED** [review]

File: examples/Inquiry.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+require_once __DIR__ . '/../vendor/autoload.php';
```
[the remark this answers is not in the document]
Comment (ar4min) [reply examples/Inquiry.php:3]: با این مدل کار نکرد..

File: src/Endpoint/GraphQL/RefundService.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            ], $body);
+
+            $this->checkHttpError($response);
+
+            $responseData = json_decode($response->getBody()->getContents(), true, 512, JSON_THROW_ON_ERROR);
+
+        } catch (JsonException $e) {
+            throw new ResponseException('JSON parsing error: ' . $e->getMessage(), -98, null, ['details' => $e->getMessage()]);
+        } catch (ResponseException $e) {
+            throw $e;
```
Comment (MA7) [inline src/Endpoint/GraphQL/RefundService.php:54]: $e->?

File: src/Endpoint/GraphQL/RefundService.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        } catch (Exception $e) {
+            throw new ResponseException('Request failed: ' . $e->getMessage(), -99, null, ['details' => $e->getMessage()]);
+        }
+
+        return $this->checkGraphQLError($responseData);
+    }
+
+    private function checkHttpError(ResponseInterface $response): void
```
Comment (MA7) [inline src/Endpoint/GraphQL/RefundService.php:62]: dupicate

File: src/Endpoint/GraphQL/TransactionService.php [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+        return $transactions;
+    }
+
+    private function httpHandler(string $uri, string $body): array
+    {
+        try {
+            $httpClient = $this->clientBuilder->getHttpClient();
+            $response = $httpClient->post($uri, [
+                'User-Agent' => sprintf('%sSdk/v.0.1 (php %s)', $this->getClassName(), PHP_VERSION),
```
Comment (MA7) [inline src/Endpoint/GraphQL/TransactionService.php:46]: make CONST

File: src/Endpoint/GraphQL/TransactionService.php [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+            $responseContent = $response->getBody()->getContents();
+            $this->checkHttpError($response);
+
+            $responseData = json_decode($responseContent, true, 512, JSON_THROW_ON_ERROR);
+
+        } catch (JsonException $e) {
+            throw new ResponseException('JSON parsing error: ' . $e->getMessage(), -98, null, ['details' => $e->getMessage()]);
+        } catch (ResponseException $e) {
+            throw $e;
+        } catch (Exception $e) {
```
Comment (MA7) [inline src/Endpoint/GraphQL/TransactionService.php:60]: } catch (Exception $e) {
            throw new ResponseException('Request failed: ' . $e->getMessage(), $e->getCode(), null, ['details' => $e->getMessage()]);
        }

File: src/Options.php [hunk 12]
```diff
@@ -24,19 +23,25 @@ public function __construct(array $options = [])
...
-                'client_builder' => new ClientBuilder(),
-                'uri_factory' => Psr17FactoryDiscovery::findUriFactory(),
-                'base_url' => $this->arrayGet(getenv(), 'ZARINPAL_MERCHANT_URL', 'https://api.zarinpal.com'),
-                'merchant_id' => $this->arrayGet(getenv(), 'ZARINPAL_MERCHANT_KEY', 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx')
-            ]
-        );
+        $resolver->setDefaults([
+            'client_builder' => new ClientBuilder(),
+            'uri_factory' => Psr17FactoryDiscovery::findUriFactory(),
+            'base_url' => $this->arrayGet(getenv(), 'ZARINPAL_BASE_URL', 'https://payment.zarinpal.com'),
```
Comment (MA7) [inline src/Options.php:29]: add .env.sample

File: src/Validator.php [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    public static function validateAuthority(string $authority): void
+    {
+        if (!preg_match('/^[AS][0-9a-zA-Z]{35}$/', $authority)) {
+            throw new InvalidArgumentException('Invalid authority format. It should be a string starting with "A" or "S" followed by 35 alphanumeric characters.');
+        }
+    }
+
+    public static function validateAmount(int $amount, int $minAmount = 1): void
```
Comment (MA7) [inline src/Validator.php:29]: 100 Toman 1000 Rial

File: src/Validator.php [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        if ($amount < $minAmount) {
+            throw new InvalidArgumentException("Amount must be at least {$minAmount}.");
+        }
+    }
+
+    public static function validateCallbackUrl(string $callback_url): void
+    {
+        if (!preg_match('/^https?:\/\/.*/', $callback_url)) {
```
Comment (MA7) [inline src/Validator.php:38]: http is valid too

**Review by MA7 — CHANGES_REQUESTED** [review]

**Pull request merged**